crates/indexer/src/models/token_models/v2_token_utils.rs

```diff
@@ -0,0 +1,434 @@
+    pub decimals: i32,
+    pub symbol: String,
+}
+
+impl FungibleAssetMetadata {
+    pub fn from_write_resource(
+        write_resource: &WriteResource,
+        txn_version: i64,
+    ) -> anyhow::Result<Option<Self>> {
+        let type_str = format!(
+            "{}::{}::{}",
+            write_resource.data.typ.address,
+            write_resource.data.typ.module,
+            write_resource.data.typ.name
+        );
+        if !V2TokenResource::is_resource_supported(type_str.as_str()) {
+            return Ok(None);
+        }
+        let resource = MoveResource::from_write_resource(
+            write_resource,
+            0, // Placeholder, this isn't used anyway
+            txn_version,
+            0, // Placeholder, this isn't used anyway
+        );
+
+        if let V2TokenResource::FungibleAssetMetadata(inner) =
+            V2TokenResource::from_resource(&type_str, resource.data.as_ref().unwrap(), txn_version)?
+        {
+            Ok(Some(inner))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct FungibleAssetStore {
+    pub metadata: ResourceReference,
+    #[serde(deserialize_with = "deserialize_from_string")]
+    pub balance: BigDecimal,
+    pub frozen: bool,
+}
+
+impl FungibleAssetStore {
+    pub fn from_write_resource(
+        write_resource: &WriteResource,
+        txn_version: i64,
+    ) -> anyhow::Result<Option<Self>> {
+        let type_str = format!(
+            "{}::{}::{}",
+            write_resource.data.typ.address,
+            write_resource.data.typ.module,
+            write_resource.data.typ.name
+        );
+        if !V2TokenResource::is_resource_supported(type_str.as_str()) {
+            return Ok(None);
+        }
+        let resource = MoveResource::from_write_resource(
+            write_resource,
+            0, // Placeholder, this isn't used anyway
+            txn_version,
+            0, // Placeholder, this isn't used anyway
+        );
+
+        if let V2TokenResource::FungibleAssetStore(inner) =
+            V2TokenResource::from_resource(&type_str, resource.data.as_ref().unwrap(), txn_version)?
+        {
+            Ok(Some(inner))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct OptionalSupply {
+    vec: Vec<Supply>,
+}
+
+impl OptionalSupply {
+    pub fn get_supply(&self) -> Option<BigDecimal> {
+        if let Some(supply) = self.vec.first() {
+            supply.get_supply()
+        } else {
+            None
+        }
+    }
+
+    pub fn get_maximum(&self) -> Option<BigDecimal> {
+        if let Some(supply) = self.vec.first() {
+            supply.get_maximum()
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BigDecimalVectorWrapper {
+    #[serde(deserialize_with = "deserialize_from_string")]
+    pub inner: BigDecimal,
+}
```

Comment:
@bowenyang007 I thought you did this before for optional aggregator.

Reply:
Yes but htis was a little different. it wasn't optional<u128>. It was optional<SomeWrapper>